### PR TITLE
KAFKA-8425: Fix for correctly handling immutable maps (KIP-421 bug)

### PR DESCRIPTION
**Pull request by tadsul**
The originals map passed to the AbstractConfig class can be immutable. The resolveConfigVariable function was modifying the originals map. If this map is immutable it will result in an exception.

This fix resolves this issue

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java
Comment (abbccdda) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: It's a bit weird to convert an immutable map to mutable and never convert back. Could we refactor the logic to push up the modification logic?

Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: @tadsul We could perhaps convert this to a immutable map and store `in `originals` like we do for `values`.

File: clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:367]: Does this test fail without the change?

**Review by rajinisivaram — COMMENTED** [review]
@tadsul Thanks for the PR. Left a couple of comments. Can we make sure that we have a test that fails without the change?

**Comment by rajinisivaram** [comment]
@tadsul There are checkstyle failures in the PR build. Could you please fix those as well?

**Comment by rajinisivaram** [comment]
@tadsul Can we create a JIRA and set 2.3.0 as fix version to make sure it gets merged into 2.3.0 before thre release?

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java
Comment (rhauch) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: I'd suggest naming this variable `resolvedOriginals`.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java
Comment (rhauch) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Then, this method should return an immutable version of the resolved originals:
```
return Collections.unmodifiableMap(resolvedOriginals);
```

File: clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java
Comment (rhauch) [reply clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:367]: @tadsul, what is the answer to this question? (Unresolving this conversation.)

Comment (tadsul) [reply clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:367]: yes the test fails without the comment

Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:367]: Hmm, I am not sure it would. You should run the test without the changes to `AbstractConfig.java` to verify. The test invokes `new TestIndirectConfigResolution(props, immutableMap)` where neither `props` nor `immutableMap` is actually immutable. In order fot the test to fail `props` needs to be unmodifiable since that is the map used as `originals`. And `immutableMap` needs to be renamed since it is confusing  as it is neither immutable nor the map being tested for immutability.

**Comment by rajinisivaram** [comment]
@tadsul At the moment in `AbstractConfig` constructor, we are doing:
```
this.originals = resolveConfigVariables(configProviderProps, (Map<String, Object>) originals);
```
This doesn't work as expected for two reasons:
1) If we resolve configs in `originals` and then invoke plugins with the `originals` that are already resolved, we resolve configs again. I think we dont want that behaviour. This was the case even before this PR.
2) As you already figured out, this PR breaks `AbstractConfigTest#testUnused` because we are replacing `RecordingMap` with a new map. We shouldn't do that either.

Can't think of a way to fix this behaviour neatly. The  only hacky solution I can think of that fixes 1) and 2) would be to check if `originals` passed through the constructor is a RecordingMap and only resolve if it is not.

2) is already tested by `AbstractConfigTest#testUnused`. We should also add a test for 1).

File: clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java
Comment (tadsul) [reply clients/src/test/java/org/apache/kafka/common/config/AbstractConfigTest.java:367]: oh yes i had to modify the test to make it work .. my yes meant that i could reproduce the issue.. sorry about that i will post the modified test

**Comment by tadsul** [comment]
AbstractConfig currently expects the following behavior 
1. When a abstract config object is instantiated it parses the values from originals and stores the values in a values map. The config values in the values map cannot be changed after instantiation 
2. All the prefix operations in abstract config rely on the 'originals' instance. So any changes made to the originals instance after the abstractConfig object is constructed should be reflected in the get operations.  
e.g. Streams instantiates the streamConfig with 'props' map, after the config is instantiated it updates the props map with new prefix values, the expected behavior is that abstractConfig should reflect these updated values when a call is made valuesWithPrefixOverride or any of the prefix methods.

In order to support the current behavior we cannot create a copy of the originals map to save the resolved configs. So this PR proposes the following
1. For Mutable Maps we save the resolved configs in the originals map itself. 
2. For Immutable maps, if there are variables in originals map we make a copy of the immutable map and save the resolved configs in the copy and make it immutable. The assumption here is that as the map is immutable its values cannot be changed after the instance is created hence it wont reflect any changes made to the originals map after the abstract config instance is created.

**Comment by tadsul** [comment]
The 'originals' map in AbstractConfig can be mutable or immutable instance. In order to resolve and save the variable configs we have added a new ResolvingMap class. The ResolvingMap class extends from HashMap and keeps a track of original map instance and the resolved configs. It stores the original input map as-is in a separate internal immutable `Map` field, and stores all  the entries (including the resolved ones) in the super. 

The 'originals' map can also be an instance of RecordingMap.  In order to ensure that all the usage are recorded correctly the ResolvingMap overrides get() function to perform the following
1) always call `originals.get(key)` to record the access.
2) always return `super.get(key)` so it it returns the resolved value to the caller.

**Comment by rhauch** [comment]
This change will require #6855 to be merged first, which addresses an issue with this PR's builds in which the Kafka Streams tests changed the originals map *after* instantiating the StreamsConfig class. Once that is merged, this PR's builds should pass.

**Review by rhauch — APPROVED** [review]
Thanks, @tadsul! LGTM, pending a green build after #6855 is merged.

**Comment by rhauch** [comment]
This will need to be backported to the `2.3` branch, which is the first branch with the config variable resolution via config providers.

**Comment by guozhangwang** [comment]
#6855 has been merged. Retest this please

**Comment by rhauch** [comment]
Retest this

**Comment by rhauch** [comment]
retest this please

**Review by rajinisivaram — APPROVED** [review]
@tadsul Thanks for the updates, LGTM. Merging to trunk and 2.3.

**Pull request merged**
